### PR TITLE
feat(erc20-price): Add chainlink aggregators

### PR DESCRIPTION
**Pull request by harsh9200**

### CODE REVIEW

**Review by this-username-is-taken — DISMISSED** [review]

File: erc20-price/src/lib.rs [hunk 1]
```diff
@@ -39,3 +48,91 @@ fn map_price_for_tokens(
...
+                .call(aggregator_address.to_vec())
+                .unwrap_or(Vec::<u8>::new());
+
+            if nested_aggregator.is_empty().not() {
+                // In `AggregatorFacade` contracts, the aggregator contract is nested two times.
+                aggregator_address = nested_aggregator;
+            }
+
+            let aggregator = Aggregator {
+                address: address_pretty(&aggregator_address).to_string(),
```
Comment (this-username-is-taken) [inline erc20-price/src/lib.rs:86]: it's unnecessary to pretty the address during storage and data passing

File: erc20-price/src/lib.rs [hunk 2]
```diff
@@ -39,3 +48,91 @@ fn map_price_for_tokens(
...
+                description: description.clone(),
+                base: base_quote[0].to_string(),
+                quote: base_quote[1].to_string(),
+                decimals: decimals.to_u64(),
+            };
+
+            output.set(
+                0,
+                format!("aggregator:{}", address_pretty(&aggregator_address)),
```
Comment (this-username-is-taken) [inline erc20-price/src/lib.rs:95]: same here

File: erc20-price/src/lib.rs [hunk 3]
```diff
@@ -39,3 +48,91 @@ fn map_price_for_tokens(
...
+
+                let erc20price = Erc20Price {
+                    block_number: block.number,
+                    price_usd: token_price.to_string(),
+                    token_address: token_address.clone(),
+                };
+
+                output.set(
+                    0,
+                    format!("chainlink_price:{}", &token_address),
```
Comment (this-username-is-taken) [inline erc20-price/src/lib.rs:132]: this should be moved to keyer for better encapsulation

File: erc20-price/src/lib.rs [hunk 4]
```diff
@@ -39,3 +48,91 @@ fn map_price_for_tokens(
...
+fn store_chainlink_price(
+    block: eth::Block,
+    store: StoreGetProto<Aggregator>,
+    output: StoreSetProto<Erc20Price>,
+) {
+    for log in block.logs() {
+        if let Some(event) = chainlink_aggregator::events::AnswerUpdated::match_and_decode(log) {
+            let aggregator_address = address_pretty(log.address());
+
+            if let Some(aggregator) = store.get_last(format!("aggregator:{}", aggregator_address)) {
```
Comment (this-username-is-taken) [inline erc20-price/src/lib.rs:112]: same here

File: erc20-price/src/lib.rs [hunk 5]
```diff
@@ -39,3 +48,91 @@ fn map_price_for_tokens(
...
+                if ["USD", "DAI", "USDC", "USDT"]
```
Comment (this-username-is-taken) [inline erc20-price/src/lib.rs:113]: nice!

File: erc20-price/src/lib.rs [hunk 6]
```diff
@@ -39,3 +48,91 @@ fn map_price_for_tokens(
...
+                    .not()
+                {
+                    // TODO: add logic for handling `ETH` quote.
+                    continue;
+                }
+
+                let token_address = aggregator.address;
+                let token_price = event.current.to_decimal(aggregator.decimals);
+
+                let erc20price = Erc20Price {
```
Comment (this-username-is-taken) [inline erc20-price/src/lib.rs:124]: @steegecs was asking for a price source. We can add it here

File: erc20-price/src/lib.rs (hunk 1 above)
Comment (harsh9200) [reply erc20-price/src/lib.rs:86]: Noted!

File: erc20-price/src/lib.rs (hunk 6 above)
Comment (harsh9200) [reply erc20-price/src/lib.rs:124]: Yes, I will add that

File: erc20-price/proto/v1/erc20_price.proto [hunk 7]
```diff
@@ -2,12 +2,13 @@ syntax = "proto3";
...
 package messari.erc20_price.v1;
 
-message ERC20Price {
-  uint64 block_number = 1;
+message Erc20Price {
+  string token_address = 1;
   string price_usd = 2;
-  string token_address = 3;
+  uint64 block_number = 3;
+  string source = 4;
```
Comment (this-username-is-taken) [inline erc20-price/proto/v1/erc20_price.proto:9]: this should be an enum

File: erc20-price/proto/v1/chainlink.proto [hunk 8]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+
+package messari.chainlink.v1;
+
+message Aggregator {
+  string address = 1;
+  string description = 2;
+  string base = 3;
+  string base_address = 4;
+  string quote = 5;
```
Comment (this-username-is-taken) [inline erc20-price/proto/v1/chainlink.proto:10]: I think we should still have quote address for the general case. Consider our use of chainlink price feed for the price module as a special use case of the chainlink aggregator, and there could be other use cases that we may not know about that could require the quote address

File: erc20-price/proto/v1/erc20_price.proto [hunk 9]
```diff
@@ -2,12 +2,13 @@ syntax = "proto3";
 
 package messari.erc20_price.v1;
 
-message ERC20Price {
-  uint64 block_number = 1;
+message Erc20Price {
+  string token_address = 1;
   string price_usd = 2;
-  string token_address = 3;
+  uint64 block_number = 3;
```
Comment (this-username-is-taken) [inline erc20-price/proto/v1/erc20_price.proto:8]: maybe a comment describing what the `block_number` refers to? I'm assuming it is the "last updated block"

File: erc20-price/src/lib.rs [hunk 10]
```diff
@@ -30,12 +40,114 @@ fn map_price_for_tokens(
                 substreams::errors::Error::Unexpected(format!("Failed to get price: {}", e))
             })?;
         prices.items.push(Erc20Price {
-            block_number,
+            block_number: block_number,
             price_usd: token_price.to_string(),
             token_address: Hex(erc20_token.clone()).to_string(),
+            source: "Oracle".to_string(),
```
Comment (this-username-is-taken) [inline erc20-price/src/lib.rs:46]: This should be an enum

File: erc20-price/src/lib.rs [hunk 11]
```diff
@@ -30,12 +40,114 @@ fn map_price_for_tokens(
...
+            }
+
+            let mut aggregator_address = decoded_call.aggregator;
+
+            let nested_aggregator = (chainlink_aggregator::functions::Aggregator {})
+                .call(aggregator_address.to_vec())
+                .unwrap_or(Vec::<u8>::new());
+
+            if nested_aggregator.is_empty().not() {
+                // In `AggregatorFacade` contracts, the aggregator contract is nested two times.
```
Comment (this-username-is-taken) [inline erc20-price/src/lib.rs:83]: would be helpful to include an example of that in the comments

File: erc20-price/proto/v1/erc20_price.proto [hunk 12]
```diff
@@ -2,12 +2,17 @@ syntax = "proto3";
...
 
-message ERC20Price {
-  uint64 block_number = 1;
+message Erc20Price {
+  string token_address = 1;
   string price_usd = 2;
-  string token_address = 3;
+  uint64 block_number = 3; // last updated block number
+  enum Source {
+    oracles = 0;
```
Comment (this-username-is-taken) [inline erc20-price/proto/v1/erc20_price.proto:10]: these should be capitalized: https://developers.google.com/protocol-buffers/docs/style#enums

**Pull request merged**